agent: backhaul_manager: Implement AP Metrics Query/Response message part 2

VladyslavTupikin opened the pull request:
The second part should contain implementation for methods for creation tlvAssociatedStaLinkMetrics and tlvAssociatedStaTrafficStats payloads and update for ap_metrics_query_message in the monitor_thread for using 2 new methods in Policy Config enabled and the same implementation in the ap_metrics_response message for the backhaul manager.

Commit:
agent: monitor: add methods for AssociatedStaTrafficStats/LinkMetrics

For creating tlvAssociatedStaTrafficStats and tlvAssociatedStaLinkMetrics
need to implement 2 methods with filling tlv data in the monitor_stats file.

Create 2 methods with names add_ap_assoc_sta_traffic_stat and
add_ap_assoc_sta_link_metric, fill the tlv data using information
from the monitor_sta_node.

add_ap_assoc_sta_link_metric is very similar to the existing
send_associated_sta_link_metrics - however, the latter uses a VS message
instead of the EasyMesh message. That needs to be refactored later.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>
Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

Commit:
agent: backhaul_manager: Add 2 tlvs in ap_metrics_response

Need to add tlvAssociatedStaLinkMetrics and tlvAssociatedStaTrafficStats
as part of AP Metrics Response message.

Add processing for tlvAssociatedStaLinkMetrics and tlvAssociatedStaTrafficStats
in the monitor thread if AP Policy Config enabled and add proccessing for those
tlv in tht ap_metrics_response for the backhaul manager.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>
Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

mariomaz commented on tests/test_flows.py:
```diff
@@ -496,34 +496,11 @@ def test_combined_infra_metrics(self):
         vap2 = env.agents[1].radios[1].vaps[0]
         env.controller.dev_send_1905(env.agents[0].mac, 0x800B,
                                      tlv(0x93, 0x0007, "0x01 {%s}" % (vap1.bssid)))
-        self.check_log(env.agents[0].radios[0], "Received AP_METRICS_QUERY_MESSAGE")
```
I think this line should not have been removed

VladyslavTupikin replied:
outdated

mariomaz commented on tests/test_flows.py:
```diff
@@ -496,34 +496,11 @@ def test_combined_infra_metrics(self):
         debug("Send AP Metrics query message to agent 2")
         env.controller.dev_send_1905(env.agents[1].mac, 0x800B,
                                      tlv(0x93, 0x0007, "0x01 {%s}" % vap2.bssid))
-        self.check_log(env.agents[1].radios[1], "Received AP_METRICS_QUERY_MESSAGE")
```
I think this line should not have been removed

VladyslavTupikin replied:
outdated.

mariomaz commented on agent/src/beerocks/monitor/monitor_stats.cpp:
```diff
@@ -492,6 +494,60 @@ bool monitor_stats::add_ap_metrics(ieee1905_1::CmduMessageTx &cmdu_tx, const sMa
+        LOG(ERROR) << "Couldn't ap_assoc_sta_link_metric_tlv->alloc_bssid_info_list";
+        return false;
+    }
+    auto stat = sta_node->get_stats();
```
const auto & stats

VladyslavTupikin replied:
solved

mariomaz commented on agent/src/beerocks/monitor/monitor_stats.cpp:
```diff
@@ -492,6 +494,60 @@ bool monitor_stats::add_ap_metrics(ieee1905_1::CmduMessageTx &cmdu_tx, const sMa
+    info.sta_measured_uplink_rssi_dbm_enc      = stat.rx_rssi_curr;
+
+    auto list         = ap_assoc_sta_link_metric_tlv->bssid_info_list(0);
+    std::get<0>(list) = true;
```
I think that the assignment of the first tuple element is useless. 

You could probably rewrite previous block as it is being done in other parts of the code:
```c
auto &info = std::get<1>(ap_assoc_sta_link_metric_tlv->bssid_info_list(0));
info.bssid = bssid;
...

VladyslavTupikin replied:
accepted this change

mariomaz commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -1435,12 +1435,28 @@ bool monitor_thread::handle_ap_metrics_query(Socket &sd, ieee1905_1::CmduMessage
-        //}
+        auto include_sta_tlv = mon_db.get_radio_node()
+                                   ->ap_metrics_reporting_info()
+                                   .include_associated_sta_traffic_stats_tlv_in_ap_metrics_response;
```
There are 2 bools in ap_metrics_reporting_info(), one for each of the optional TLVs:
```c
        /**
         * Associated STA Traffic Stats Inclusion Policy.
         * 0: Do not include Associated STA Traffic Stats TLV in AP Metrics Response
         * 1: Include Associated STA Traffic Stats TLV in AP Metrics Response
         * (Value is obtained from Metric Reporting Policy TLV)
         */
        bool include_associated_sta_link_metrics_tlv_in_ap_metrics_response = false;

        /**
         * Associated STA Link Metrics Inclusion Policy.
         * 0: Do not include Associated STA Link Metrics TLV in AP Metrics Response
         * 1: Include Associated STA Link Metrics TLV in AP Metrics Response
         * (Value is obtained from Metric Reporting Policy TLV)
         */
        bool include_associated_sta_traffic_stats_tlv_in_ap_metrics_response = false;

mariomaz commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -1435,12 +1435,28 @@ bool monitor_thread::handle_ap_metrics_query(Socket &sd, ieee1905_1::CmduMessage
+                    continue;
+                }
+
+                if (!mon_stats.add_ap_assoc_sta_traffic_stat(cmdu_tx, sta_node)) {
```
This TLV should be added if and only if include_associated_sta_link_metrics_tlv_in_ap_metrics_response == true

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2535,9 +2535,37 @@ bool backhaul_manager::handle_slave_ap_metrics_response(ieee1905_1::CmduMessageR
     }
+    std::vector<sStaTrafficStats> traffic_stats_response;
+
+    for (auto sta_traffic : cmdu_rx.getClassList<wfa_map::tlvAssociatedStaTrafficStats>()) {
```
Use `const auto &` or `auto &`

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2535,9 +2535,37 @@ bool backhaul_manager::handle_slave_ap_metrics_response(ieee1905_1::CmduMessageR
+    }
+
+    std::vector<sStaApLinkMetrics> link_metrics_response;
+    for (auto sta_link_metric : cmdu_rx.getClassList<wfa_map::tlvAssociatedStaLinkMetrics>()) {
```
Use const auto & or auto &

mariomaz replied:
According to [CONTRIBUTING.md](https://github.com/prplfoundation/prplMesh/blob/master/CONTRIBUTING.md), you should mark the comments that have been addressed as Resolved only if they have been resolved. If you don't take the suggestion directly or you disagree or only apply it in part, don't mark it as Resolved.

arnout replied:
Something weird happened here:

8 days ago, @mariomaz wrote:
```
for (auto sta_link_metric : cmdu_rx.getClassList<wfa_map::tlvAssociatedStaLinkMetrics>()) {
```
> Use const auto & or auto &

Yesterday, I pushed:
```
for (auto &sta_link_metric : cmdu_rx.getClassList<wfa_map::tlvAssociatedStaLinkMetrics>()) {
```

and Mario wrote:
> If you don't take the suggestion directly or you disagree or only apply it in part, don't mark it as Resolved.

I guess Mario somehow reviewed the old version?

arnout replied:
Ah, hang on... I usually mark things as resolved while editing, so before it actually gets pushed. So most likely Mario wrote this before I pushed my update. Sorry for the noise.

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2580,6 +2608,47 @@ bool backhaul_manager::handle_slave_ap_metrics_response(ieee1905_1::CmduMessageR
                     ap_metrics_response_tlv->estimated_service_info_field());
+
+        if (!response.sta_traffic_stats.empty()) {
+            for (auto stat : response.sta_traffic_stats) {
```
const auto &

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2580,6 +2608,47 @@ bool backhaul_manager::handle_slave_ap_metrics_response(ieee1905_1::CmduMessageR
                     response.metric.estimated_service_info_field.size(),
                     ap_metrics_response_tlv->estimated_service_info_field());
+
+        if (!response.sta_traffic_stats.empty()) {
```
No need to check if container is empty

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2580,6 +2608,47 @@ bool backhaul_manager::handle_slave_ap_metrics_response(ieee1905_1::CmduMessageR
+            }
+        }
+
+        if (!response.sta_link_metrics.empty()) {
```
No need to check if container is empty

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2580,6 +2608,47 @@ bool backhaul_manager::handle_slave_ap_metrics_response(ieee1905_1::CmduMessageR
+        }
+
+        if (!response.sta_link_metrics.empty()) {
+            for (auto link_metric : response.sta_link_metrics) {
```
const auto &

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2580,6 +2608,47 @@ bool backhaul_manager::handle_slave_ap_metrics_response(ieee1905_1::CmduMessageR
+                    cmdu_tx.addClass<wfa_map::tlvAssociatedStaTrafficStats>();
+
+                if (!sta_traffic_response_tlv) {
+                    LOG(ERROR) << "Failed addClass<wfa_map::tlvAssociatedStaTrafficStats>";
```
continue or break on error

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2580,6 +2608,47 @@ bool backhaul_manager::handle_slave_ap_metrics_response(ieee1905_1::CmduMessageR
+                    cmdu_tx.addClass<wfa_map::tlvAssociatedStaLinkMetrics>();
+
+                if (!sta_link_metric_response_tlv) {
+                    LOG(ERROR) << "Failed addClass<wfa_map::tlvAssociatedStaLinkMetrics>";
```
continue or break on error

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2580,6 +2608,47 @@ bool backhaul_manager::handle_slave_ap_metrics_response(ieee1905_1::CmduMessageR
+
+                sta_link_metric_response_tlv->sta_mac() = link_metric.sta_mac;
+                if (!sta_link_metric_response_tlv->alloc_bssid_info_list(1)) {
+                    LOG(ERROR) << "Failed alloc_bssid_info_list";
```
continue or break on error

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2580,6 +2608,47 @@ bool backhaul_manager::handle_slave_ap_metrics_response(ieee1905_1::CmduMessageR
+                    LOG(ERROR) << "Failed alloc_bssid_info_list";
+                }
+                auto sta_link_metric_response = sta_link_metric_response_tlv->bssid_info_list(0);
+                std::get<0>(sta_link_metric_response) = true;
```
I think that the assignment of the first tuple element is useless.

VladyslavTupikin replied:
fixed

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2580,6 +2608,47 @@ bool backhaul_manager::handle_slave_ap_metrics_response(ieee1905_1::CmduMessageR
+                }
+
+                sta_link_metric_response_tlv->sta_mac() = link_metric.sta_mac;
+                if (!sta_link_metric_response_tlv->alloc_bssid_info_list(1)) {
```
Maybe you should copy as many entries as there are in the source TLV to the destination TLV (instead of supposing that there is just one) as it is being done in some other parts of the code where a `bssid_info_list` is copied.

VladyslavTupikin replied:
No, we have just 1 bssid_info_list.

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2535,9 +2535,37 @@ bool backhaul_manager::handle_slave_ap_metrics_response(ieee1905_1::CmduMessageR
 
     // Fill a response vector
-    m_ap_metric_response.push_back({metric});
+    m_ap_metric_response.push_back({metric, traffic_stats_response, link_metrics_response});
```
Maybe it's simpler and easier to maintain if you create and store the TLVs instead of intermediate structures

VladyslavTupikin replied:
Can we create tlv without cmdu_rx/tx objects?

mariomaz replied:
cmdu_tx is defined in base class `beerocks::btl::socket_thread`, but you could define your own message here and fill it (instead of using those intermediate structures).

VladyslavTupikin replied:
It will be done as part of https://github.com/prplfoundation/prplMesh/issues/1328

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:
```diff
@@ -509,6 +510,22 @@ class backhaul_manager : public btl::transport_socket_thread {
                           const sLinkNeighbor &link_neighbor, const sLinkMetrics &link_metrics,
                           ieee1905_1::eLinkMetricsType link_metrics_type);
 
+    struct sStaTrafficStats {
```
These structs wouldn't be necessary if you create and store the TLVs to send

mariomaz reviewed (changes requested):
Some comments need to be addressed

mariomaz commented on agent/src/beerocks/monitor/monitor_stats.cpp:
```diff
@@ -492,6 +494,60 @@ bool monitor_stats::add_ap_metrics(ieee1905_1::CmduMessageTx &cmdu_tx, const sMa
 }
 
+bool monitor_stats::add_ap_assoc_sta_traffic_stat(ieee1905_1::CmduMessageTx &cmdu_tx,
+                                                  monitor_sta_node *sta_node)
```
```suggestion
                                                  const monitor_sta_node &sta_node)
```

mariomaz commented on agent/src/beerocks/monitor/monitor_stats.cpp:
```diff
@@ -492,6 +494,60 @@ bool monitor_stats::add_ap_metrics(ieee1905_1::CmduMessageTx &cmdu_tx, const sMa
+}
+
+bool monitor_stats::add_ap_assoc_sta_link_metric(ieee1905_1::CmduMessageTx &cmdu_tx,
+                                                 const sMacAddr &bssid, monitor_sta_node *sta_node)
```
```suggestion
                                                 const sMacAddr &bssid, const monitor_sta_node &sta_node)
```

arnout commented on agent/src/beerocks/monitor/monitor_stats.cpp:
```diff
@@ -492,6 +494,60 @@ bool monitor_stats::add_ap_metrics(ieee1905_1::CmduMessageTx &cmdu_tx, const sMa
+    return true;
+}
+
+bool monitor_stats::add_ap_assoc_sta_link_metric(ieee1905_1::CmduMessageTx &cmdu_tx,
```
The equivalent of this was also done by Vitalii in #1139, so we have to find out how they interact...

vitalii-komisarenko replied:
@arnout you probably wanted to say #1149.

Commit:
tests: test_flow.py: Fix combined_infra_metrics test

Need to fix missing tlvs, add checks for STA Traffic Stats and
STA Link Mettrcis.

Added checks for STA Traffic Stats and STA Link Metrics, update
FIXME: need to fix combined infra metrics test.

Associate a STA to each VAP so there actually are STA stats/metrics.

Combined infra metrics should *not* contain the STA stats/metrics, so
add a comment to make this clear.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>
Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

Commit:
Add 4.7.4_ETH_* tests to the list of passing tests.

MAP-4.7.4_ETH_FH24G:netgear-rax40
MAP-4.7.4_ETH_FH5GL:netgear-rax40
MAP-4.7.4_ETH_FH5GH:netgear-rax40

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2567,22 +2533,44 @@ bool backhaul_manager::handle_slave_ap_metrics_response(ieee1905_1::CmduMessageR
     auto info                           = ap_metrics_tlv->estimated_service_info_field();
     if (ap_metrics_tlv->estimated_service_info_field_length()) {
```
Checking the length is redundant (If 0, the loop is empty)

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2625,8 +2613,49 @@ bool backhaul_manager::handle_slave_ap_metrics_response(ieee1905_1::CmduMessageR
+                cmdu_tx.addClass<wfa_map::tlvAssociatedStaTrafficStats>();
+
+            if (!sta_traffic_response_tlv) {
+                LOG(ERROR) << "Failed addClass<wfa_map::tlvAssociatedStaTrafficStats>";
```
Loop should break or return in case of error

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2625,8 +2613,49 @@ bool backhaul_manager::handle_slave_ap_metrics_response(ieee1905_1::CmduMessageR
+                cmdu_tx.addClass<wfa_map::tlvAssociatedStaLinkMetrics>();
+
+            if (!sta_link_metric_response_tlv) {
+                LOG(ERROR) << "Failed addClass<wfa_map::tlvAssociatedStaLinkMetrics>";
```
Loop should break or return in case of error

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2625,8 +2613,49 @@ bool backhaul_manager::handle_slave_ap_metrics_response(ieee1905_1::CmduMessageR
+
+            sta_link_metric_response_tlv->sta_mac() = link_metric.sta_mac;
+            if (!sta_link_metric_response_tlv->alloc_bssid_info_list(1)) {
+                LOG(ERROR) << "Failed alloc_bssid_info_list";
```
Loop should break or return in case of error

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2625,8 +2613,49 @@ bool backhaul_manager::handle_slave_ap_metrics_response(ieee1905_1::CmduMessageR
     }
 
+    //Clear the m_ap_metric_response vector after preparing response for the controller
+    m_ap_metric_response.clear();
```
m_ap_metric_query is storing the slaves and the BSSIDs for which AP metrics have been requested, but it is not storing the MID.
When slaves reply, items in m_ap_metric_query are removed by BSSID, without considering MID. 
**Only after m_ap_metric_query is empty** is that AP Metrics Response message is sent to controller.

This mechanism might fail if agent receives an AP Metrics Query message before it has processed previous one and the BSSIDs in both messages do not match.

arnout replied:
We kind of decided to keep it as is for M1 and refactor later -> #1328

arnout commented:
Current status: test_flows fails, need to debug what is wrong.

mariomaz reviewed (changes requested):
Some small implementation issues and one design issue (I think the processing might fail if agent receives a message before it has processed previous one).

arnout committed:
agent: backhaul: AP Policy Config: do proper swapping when forwarding

The AP Polic Config Request is forwarded to the slaves. Before
forwarding, the CMDU has to be swapped to network byte order.
However, since this is done in the middle of a loop which still uses the
contents of that CMDU, it has to be swapped back to host byte order
after forwarding.

Without this, the second radio will get a swapped CMDU because it has
been swapped twice.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

RanRegev commented on framework/tlvf/src/src/CmduMessageRx.cpp:
```diff
@@ -33,6 +33,8 @@
 #include <tlvf/wfa_map/tlvApMetrics.h>
 #include <tlvf/wfa_map/tlvApRadioBasicCapabilities.h>
 #include <tlvf/wfa_map/tlvApRadioIdentifier.h>
+#include <tlvf/wfa_map/tlvAssociatedStaLinkMetrics.h>
```
should be added in a separate commit (hereinafter `commit X`)

RanRegev commented on framework/tlvf/src/src/CmduMessageRx.cpp:
```diff
@@ -208,6 +210,9 @@ std::shared_ptr<BaseClass> CmduMessageRx::parseNextTlv()
         return msg.addClass<wfa_map::tlvStaMacAddressType>();
     }
+    case (150): {
+        return msg.addClass<wfa_map::tlvAssociatedStaLinkMetrics>();
```
`commit X`

RanRegev commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2216,6 +2216,7 @@ bool backhaul_manager::handle_multi_ap_policy_config_request(ieee1905_1::CmduMes
                 if (!message_com::forward_cmdu_to_uds(radio->slave, cmdu_rx, length)) {
                     LOG(ERROR) << "Failed to forward message to slave " << radio->radio_mac;
                 }
+                cmdu_rx.swap(); // swap back to normal after forwarding, for next iteration
```
Probably this is the correct thing to do _now and here_, however this is a bad mechanism in few ways, and it should be changed:
* at this point in the code the developer should consider _nothing_ regarding byte order, this place is to apply logic, not mechanics.
* even if byte order should be considered in this place the function must be named correctly and explicitly: `set_network_order()` and `set_host_order` for example.

arnout replied:
Absolutely, the whole idea of needing to keep track of swapping is insane.

RanRegev commented on agent/src/beerocks/monitor/monitor_stats.cpp:
```diff
@@ -457,6 +460,58 @@ bool monitor_stats::add_ap_metrics(ieee1905_1::CmduMessageTx &cmdu_tx, const sMa
+        LOG(ERROR) << "Couldn't addClass tlvAssociatedStaTrafficStats";
+        return false;
+    }
+    auto stat                                        = sta_node.get_stats().hal_stats;
```
This area fits exactly to the `gate::load` idiom.

RanRegev commented on agent/src/beerocks/monitor/monitor_stats.cpp:
```diff
@@ -457,6 +460,58 @@ bool monitor_stats::add_ap_metrics(ieee1905_1::CmduMessageTx &cmdu_tx, const sMa
 }
 
+bool monitor_stats::add_ap_assoc_sta_traffic_stat(ieee1905_1::CmduMessageTx &cmdu_tx,
+                                                  monitor_sta_node &sta_node)
```
sta_node appears only on the rhs, so probably it should be `const`
```suggestion
                                                  const monitor_sta_node &sta_node)
```

RanRegev commented on agent/src/beerocks/monitor/monitor_stats.cpp:
```diff
@@ -457,6 +460,58 @@ bool monitor_stats::add_ap_metrics(ieee1905_1::CmduMessageTx &cmdu_tx, const sMa
+
+    ap_assoc_sta_link_metric_tlv->sta_mac() = tlvf::mac_from_string(sta_node.get_mac());
+
+    // there is exactly 1 BSSID per STA
```
I can't understand how this comment supports the code below it.
maybe:
```suggestion
    // default allocation creates a list with one element which it what we want: exactly 1 BSSID per STA
```

RanRegev commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -1498,21 +1498,57 @@ bool monitor_thread::handle_ap_metrics_query(Socket &sd, ieee1905_1::CmduMessage
+        }
+
         if (!mon_stats.add_ap_metrics(cmdu_tx, bssid)) {
             return false;
```
Missing an error log message (old code but yet)

arnout replied:
The add_ap_metrics function itself already prints error logs.

RanRegev commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -1498,21 +1498,57 @@ bool monitor_thread::handle_ap_metrics_query(Socket &sd, ieee1905_1::CmduMessage
+
+        if (include_sta_traffic_stats_tlv || include_sta_link_metrics_tlv) {
+            for (auto it = mon_db.sta_begin(); it != mon_db.sta_end(); it++) {
+                auto sta_mac  = it->first;
```
```suggestion
                const auto sta_mac  = it->first;
```

arnout replied:
Actually:
```suggestion
                const auto &sta_mac  = it->first;
```
:-)

RanRegev commented on agent/src/beerocks/monitor/monitor_thread.cpp:
```diff
@@ -1498,21 +1498,57 @@ bool monitor_thread::handle_ap_metrics_query(Socket &sd, ieee1905_1::CmduMessage
+        if (include_sta_traffic_stats_tlv || include_sta_link_metrics_tlv) {
+            for (auto it = mon_db.sta_begin(); it != mon_db.sta_end(); it++) {
+                auto sta_mac  = it->first;
+                auto sta_node = it->second;
```
```suggestion
                const auto sta_node = it->second;
```

RanRegev commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2537,9 +2537,37 @@ bool backhaul_manager::handle_slave_ap_metrics_response(ieee1905_1::CmduMessageR
     }
+    std::vector<sStaTrafficStats> traffic_stats_response;
+
+    for (auto &sta_traffic : cmdu_rx.getClassList<wfa_map::tlvAssociatedStaTrafficStats>()) {
```
this line is not clear. I think it should be split:
```suggestion
    auto &associated_traffic_stats_list = cmdu_rx.getClassList<wfa_map::tlvAssociatedStaTrafficStats>()
    for (auto &sta_traffic : associated_traffic_stats_list ) {
```

arnout replied:
I disagree that this is not clear - by introducing the extra variable, I first have to learn what its meaning is and then that it's used in the loop, while the current line literally says "loop sta_traffic over cmdu_rx's TLVs of type tlvAssociatedStaTrafficStats". Well, with the minor naming mistake that it's getClassList instead of the more descriptive and coding-style-conforming get_tlv_list or get_tlvs.

RanRegev commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2582,10 +2610,50 @@ bool backhaul_manager::handle_slave_ap_metrics_response(ieee1905_1::CmduMessageR
 
-    LOG(DEBUG) << "Sending AP_METRICS_RESPONSE_MESSAGE, mid=" << std::hex << int(mid);
+    // Clear the m_ap_metric_response vector after preparing response to the controller
+    m_ap_metric_response.clear();
```
In my previous review to this file I commented on line [2578](https://github.com/prplfoundation/prplMesh/pull/1285/commits/bfae2654401e77d0b7928c5b716dc07828980f90#diff-80c48bbadac679bad34f8ee37e30f772R2578) that might never be true in case we didn't receive the result even for one bssid.
As far as I can see, in case of not receiving all responses - not only that this specific metric query will fail, all subsequent queries will fail too as `m_ap_metric_query` will never be cleared.

arnout replied:
#1328

RanRegev commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2582,10 +2610,50 @@ bool backhaul_manager::handle_slave_ap_metrics_response(ieee1905_1::CmduMessageR
+    // Clear the m_ap_metric_response vector after preparing response to the controller
+    m_ap_metric_response.clear();
 
+    LOG(DEBUG) << "Sending AP_METRICS_RESPONSE_MESSAGE, mid=" << std::hex << int(mid);
```
good! beacon in the log

Commit:
framework: tlvf: Add AssociatedStaTrafficStats TLV

The yaml file exists already, it just needs to be enabled in
tlvf_conf.yaml.

Update AutoGenerated.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>
Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
framework: CmduMessageRx: add parse for AssociatedStaTrafficStats/LinkMetrics

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>
Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
agent: monitor_db: monitor_sta_node: constify getter functions

Getter functions should be const because they don't modify the object.
This allows us to pass monitor_sta_node objects by const-reference as
our coding style prescribes.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

mariomaz commented on tests/test_flows.py:
```diff
@@ -633,11 +633,23 @@ def test_link_metric_query(self):
+
+        time.sleep(1)
+
+        debug("Send AP Metrics query message to agent 1 with Traffic Stats for {}".format(sta1.mac))
```
Debug message is mixing AP Metrics Query message with Multi-AP Policy Config Request message

VladyslavTupikin replied:
Fixed

mariomaz commented on tests/test_flows.py:
```diff
@@ -646,13 +658,20 @@ def test_combined_infra_metrics(self):
                                                env.controller.mac, mid)
         debug("Check AP metrics response has AP metrics")
```
```suggestion
        debug("Check AP Metrics Response message has AP Metrics TLV")
```

VladyslavTupikin replied:
fixed

mariomaz commented on tests/test_flows.py:
```diff
@@ -646,13 +658,20 @@ def test_combined_infra_metrics(self):
+
+        debug("Check AP metrics response has STA traffic stats")
+        sta_stats_1 = self.check_cmdu_has_tlv_single(response, 0xa2)
+        if sta_stats_1:
```
Test should check that response does not contain any Associated STA Link Metrics TLV

VladyslavTupikin replied:
Created issue https://github.com/prplfoundation/prplMesh/issues/1340

mariomaz commented on tests/test_flows.py:
```diff
@@ -661,9 +680,22 @@ def test_combined_infra_metrics(self):
+                    ap_metrics_2.ap_metrics_bssid, vap2.bssid))
+
+        debug("Check AP metrics response has STA Link Metrics")
+        sta_metrics_2 = self.check_cmdu_has_tlv_single(response, 0x96)
```
Test should check that response does not contain any Associated STA Traffic Stats TLV

VladyslavTupikin replied:
https://github.com/prplfoundation/prplMesh/issues/1340

mariomaz commented on tests/test_flows.py:
```diff
@@ -646,13 +658,20 @@ def test_combined_infra_metrics(self):
+                self.fail("STA traffic stats with wrong MAC {} instead of {}".format(
+                    sta_stats_1.assoc_sta_traffic_stats_mac_addr, sta1.mac))
+
+        debug("Send AP Metrics query message to agent 2 with STA Metrics for {}".format(sta2.mac))
```
Debug message is mixing AP Metrics Query message with Multi-AP Policy Config Request message

VladyslavTupikin replied:
fixed

mariomaz reviewed (approved):
Just a few nitpick comments, so approving

arnout commented:
@mergifyio rebase
